docs/decision_record.md

```diff
@@ -6,6 +6,55 @@
 
 [TOC]
 
+## 2022-06-02 - How array elements are separated and digested 
+
+### Decision
+
+#### For converting from level 2 to level 1
+
+Each array of data will be digested as string of comma separated values.
+For example the array at level 2:
+```
+"lengths": [
+    248956422,
+    242193529,
+    198295559
+  ]
+```
+will be digested as: 
+```
+248956422,242193529,198295559
+```
+
+#### Conversion from level 1 to level 0
+
+For each array name sorted alphabetically in ascending order, the array name and the digest list calculated above will be listed separated by commas. 
+For Example the collection at level 1: 
+```
+{ 
+    'sequences': '8dd93796fa0225e92eb159a8779f1b254776557f748f8bfb',
+    'lengths': '501fd98e2fdcc276c47306bd72c9155489ed2b23123ddfa2',
+    'names': '7bc90a07cf25f2f64f33baee3d420ad1ae5f442055280d43',
+}
+```
+will be digested as: 
+```
+lengths,501fd98e2fdcc276c47306bd72c9155489ed2b23123ddfa2,names,7bc90a07cf25f2f64f33baee3d420ad1ae5f442055280d43,sequences,8dd93796fa0225e92eb159a8779f1b254776557f748f8bfb
+```
+
+### Rationale
+
+Following on the decision to use an [array-based data structure and multi-tiered digests](docs/decision_record.md#2021-06-30---use-array-based-data-structure-and-multi-tiered-digests), we only need to define one field separator. Comma is a natural separataor and is already part of the [restricted character set](docs/decision_record.md#2021-06-30---use-the-sam-specification-v1-description-of-sequence-names) for sequence name in the SAM specification.
+
+### Linked issues
+
+ - [https://github.com/ga4gh/seqcol-spec/issues/1](https://github.com/ga4gh/seqcol-spec/issues/1)
+ - [https://github.com/ga4gh/seqcol-spec/issues/25](https://github.com/ga4gh/seqcol-spec/issues/25)
+
+### Known limitations
+
+Using comma as a delimiter will limit the ability to have comma in the level 2 values and in the array names to avoid confusion. 
```

Comment:
I have some issues with this part. I think it is likely that someone would like to have commas in the level 2 values at some point. One possible use case that comes to mind is a need to encode genome regions as an array value, e.g. "(123,234),(234,345)".

Looking at the digest serialization algorithm specified in the [GA4GH Variation Representation Specification](https://vrs.ga4gh.org/en/latest/impl-guide/computed_identifiers.html#digest-serialization), I see that following that algorithm will remove the limitation on comma characters with minimal overhead. The unique digest of the example content above would then become:

```
{"lengths":"501fd98e2fdcc276c47306bd72c9155489ed2b23123ddfa2","names":"7bc90a07cf25f2f64f33baee3d420ad1ae5f442055280d43","sequences":"8dd93796fa0225e92eb159a8779f1b254776557f748f8bfb"}
```

This will also be a valid JSON serialization (but not all valid JSON serializations would be a valid digest serialization). Why not copy this?

Reply:
VRS uses four normalisation steps to ensure there is consistency between implementations. From your example too of the `(123,234)` made me think of graph genomes and a reason to encode paths into arrays. I found the [pangenome naming scheme](https://github.com/pangenome/PanSN-spec) which doesn't use commas.

Reply:
for our uses, we wouldn't benefit from quotes on array names, or brackets. I think quotes on array contents are what would enable you to use commas in array contents, so you could do:

```
lengths,"501fd98e2fdcc276c47306bd72c9155489ed2b23123ddfa2",names,"7bc90a07cf25f2f64f33baee3d420ad1ae5f442055280d43",sequences,"8dd93796fa0225e92eb159a8779f1b254776557f748f8bfb"
```

But either way, if we rely on quotes, then doesn't this make it so that quotes would not be allowed in array contents?

Reply:
Unfortunately due to summer holidays, I missed the meeting today (even though I had planned to attend). I had also planned to provide some comments prior to the meeting, which I also failed at doing. I have however now managed to write down some thoughts, and a new idea:

I see in the minutes that it has been decided to go for the simple comma and comma escape solution. I was also leaning towards that after last meeting, but I now believe this is too simple a solution due to the fact that we want to support UTF-8. UTF-8 support creates a need for us to specify exactly how UTF-8 strings should be serialized (in any case), as well as sorted (given that we allow UTF-8 for array names). Leaving such issues as undefined is not a good idea, IMHO. However, an attempt from our part to define a solution to such issues will easily lead us into a rabbit hole (and it has already partly done so). So in conclusion, I believe it will be beneficial for us to lean towards an existing standard. Which was my reason for my suggestion to follow the JSON digest scheme from [GA4GH Variation Representation Specification](https://vrs.ga4gh.org/en/latest/impl-guide/computed_identifiers.html#digest-serialization).

The GA4GH VRS pages states:

> Although several proposals exist for serializing arbitrary data in a consistent manner ([[Gibson]](https://vrs.ga4gh.org/en/latest/impl-guide/computed_identifiers.html#gibson), [[OLPC]](https://vrs.ga4gh.org/en/latest/impl-guide/computed_identifiers.html#olpc), [[JCS]](https://vrs.ga4gh.org/en/latest/impl-guide/computed_identifiers.html#jcs)), none have been ratified. As a result, [VRS](https://github.com/ga4gh/vrs/) defines a custom serialization format that is consistent with these proposals but does not rely on them for definition; it is hoped that a future ratified standard will be forward compatible with the process described here.

However, even though the serialization format is customly defined, the [Python implementation of GA4GH VRS](https://github.com/ga4gh/vrs-python/blob/main/src/ga4gh/core/_internal/identifiers.py#L189) makes use of the python library [canonicaljson](https://github.com/matrix-org/python-canonicaljson) which is following [yet another canonical JSON specification](https://spec.matrix.org/v1.3/appendices/#canonical-json), however the VRS spec does not even point to this.

So I think the GA4GH VRS solution is a bit of a mess. More generally, it seems the field of canonical JSON solutions as a whole is also a bit of a mess.

However, after last meeting, I came over the [RFC 8785
JSON Canonicalization Scheme (JCS)](https://www.rfc-editor.org/rfc/rfc8785.html), which seems to have appeared out of the canonical JSON implementation mess in 2020 as a very well thought out suggestion that has been through a lot of reviews and changes, much of which are related to technical considerations on JSON implementation internals and which are way above my head. This is a scheme similar to the custom one specified in the [GA4GH VRS](https://vrs.ga4gh.org/en/latest/impl-guide/computed_identifiers.html#digest-serialization), however it is more clearly defined and thorough. (Btw: the last link on the VRS page, as quoted above, seems to be a draft version of RFC 8785). Even though the [JSON LD 1.1 specification](https://www.w3.org/TR/json-ld11/#the-rdf-json-datatype) calls it an "emerging standard", it is not on the IETF Standards Track, but there still seems to be quite a bit of interest in adopting it here and there. Libraries are available for at least Javascript, Java, Go, .Net/C#, and Python.

One detail is that the RFC8785 does not specify any UTF normalization scheme, as it is out of the scope of the specification and should be handled at another level. Which I think makes sense.

My suggestion is then to adopt the RFC8785 scheme for the seqcol digests. This would easily solve all our digest needs, as well as having a technically detailed and properly reviewed document to refer to, as well as emerging library support. This would properly solve the issues of UTF-8 serialization (needed) and sorting (possibly needed), which we otherwise would need to handle. As a bonus, we will have JSON compatibility, which might be useful (but it is not the main argument). Lastly, it would simplify the seqcol specification.

Reply:
I've changed the ADR to reflect the decision to adopt RFC-8785. Let me know if that makes sense.